### ISSUE TITLE
matmul requires batch dimensions to match exactly; no broadcasting of the batch dim

### ISSUE THREAD

**Issue by vyncint** [vyncint/oxmera#22]
## Summary

Batched `matmul` requires the batch extents to be equal. Broadcasting a batch of 1 against a batch of N — the standard NumPy/PyTorch behaviour, and the usual way to apply one weight matrix to a batch of inputs — is rejected.

This is an **enhancement, not a bug**: the docstring is accurate about what it does ("rank-2 `[m, k] x [k, n]`, or batched rank-3 `[b, m, k] x [b, k, n]`"). Filing it because the restriction is invisible until you hit it and the error reads like a user mistake.

## Reproduction

```rust
let a = Tensor::from_slice(&data, [2, 2, 3]).unwrap();   // batch of 2
let b = Tensor::from_slice(&w,    [1, 3, 2]).unwrap();   // one shared matrix
a.matmul(&b).unwrap();
```

```
Err(ShapeMismatch { expected: Shape([2, 3, 2]), got: Shape([1, 3, 2]), op: "matmul" })
```

NumPy and PyTorch both give `[2, 2, 2]` here.

## Why it is worth having

Applying one weight matrix across a batch is the single most common batched-matmul shape in a neural network. Without broadcasting the caller must materialise the operand:

```rust
let b = b.broadcast_to(Shape::from([2, 3, 2]))?.contiguous()?;   // copies b batch-many times
```

which allocates `batch × k × n` floats to express something that needs none — the strided path already supports a zero stride, so the machinery exists.

## Scope

- Broadcast the batch dims of rank-3 operands per the standard rules (`1` broadcasts against `N`; unequal non-1 extents stay an error).
- Ideally accept rank-2 × rank-3 as well (`[m,k] @ [b,k,n]` and `[b,m,k] @ [k,n]`), which is the same feature seen from the other side.
- Update the `matmul` docstring and the ARCHITECTURE description of batched matmul once the semantics change.
- The autograd VJP needs the matching reduction: gradients for a broadcast operand must sum over the broadcast batch axis, exactly as `reduce_to_shape` already does for elementwise broadcasting.
